### PR TITLE
Add `downsampleCloudMsg` function to `noetic-devel`

### PR DESCRIPTION
**Pull request by h-wata**
## PR Type

- [x] Feature
- [ ] Bug fix
- [ ] Refactor
- [ ] Documentation
- [ ] Other

## Overview
@Alpaca-zip  mentioned that there is a plan to add the `downsampleCloudMsg` feature to `noetic-devel`, as mentioned in the comment below,  which is why this function is being added.
- https://github.com/Alpaca-zip/ultralytics_ros/issues/57#issuecomment-1929646310

## Detail
- 

## Test
- [x] I have tested this change with gazebo

## Attention
- There's **no rush**, so please review when you have time. Feel free to point out any issues with the implementation.

### CODE REVIEW

**Review by Alpaca-zip — CHANGES_REQUESTED** [review]
Thank you for submitting a Pull Request for my upcoming tasks. I appreciate your contribution and will proceed with a code review for this PR.

- The function `msg2TransformedCloud()` appears to be unused and therefore seems unnecessary. You can find it here:
[Link to code snippet](https://github.com/h-wata/ultralytics_ros/blob/d18a16aa0e2a048585c6c82d5ae3ce6923ad5d0d/src/tracker_with_cloud_node.cpp#L155-L174)

- The usage of `cloud2TransformedCloud()` is correct in the following part of the code:
[Link to code snippet](https://github.com/h-wata/ultralytics_ros/blob/d18a16aa0e2a048585c6c82d5ae3ce6923ad5d0d/src/tracker_with_cloud_node.cpp#L95)
However, the way `cloud2TransformedCloud()` is utilized here seems to be incorrect:
[Link to code snippet](https://github.com/h-wata/ultralytics_ros/blob/d18a16aa0e2a048585c6c82d5ae3ce6923ad5d0d/src/tracker_with_cloud_node.cpp#L57)
It should actually contain the following transformation:
  ```cpp
  geometry_msgs::TransformStamped tf =
          tf_buffer_->lookupTransform(cam_model_.tfFrame(), header.frame_id, header.stamp);
  ```
  Therefore, it would be beneficial to modify `cloud2TransformedCloud()` to take the source and target frames as arguments to increase its versatility.

- Regarding the GitHub Actions, all checks are failing due to an oversight on my part. It seems that removing the following lines should fix the issue:
[Link to CI configuration](https://github.com/h-wata/ultralytics_ros/blob/d18a16aa0e2a048585c6c82d5ae3ce6923ad5d0d/.github/workflows/noetic-ci.yml#L11-L12)
  This fix is beyond the scope of this PR, so I will address it in a separate PR. However, if it's of concern to you, feel free to make the correction within this PR.

**Comment by Alpaca-zip** [comment]
I have checked the changes made so far. I would appreciate it if you could confirm the following points.

- In the `cloud2TransformedCloud()` method described here:  https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/src/tracker_with_cloud_node.cpp#L57  it is necessary to transform the point cloud from `header.frame_id` (the LiDAR frame) to `cam_model_.tfFrame()` (the camera frame).

  On the other hand, in the `cloud2TransformedCloud()` method described here:  https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/src/tracker_with_cloud_node.cpp#L96  it is necessary to transform the point cloud from `cam_model_.tfFrame()` (the camera frame) to `header.frame_id` (the LiDAR frame).

  Currently, the implementation of `cloud2TransformedCloud()` in both cases is transforming the pointcloud from `cam_model_.tfFrame()` (the camera frame) to `header.frame_id` (the LiDAR frame).

  Here is an example implementation of `cloud2TransformedCloud()` that you might find useful:  https://github.com/Alpaca-zip/ultralytics_ros/blob/a8a6a7e4d4a9a2609148f9dc8a6eed45d2ce9804/src/tracker_with_cloud_node.cpp#L245-L263

  **Notice**
  Please be aware that the first argument of `lookupTransform()` is `target_frame`, and the second argument is `source_frame`.
  [Link to reference](https://docs.ros2.org/foxy/api/tf2_ros/classtf2__ros_1_1Buffer.html#a3ab502cc1e8b608957a96ad350815aee)

- The following section means explicitly specifying the source branch of the pull request for checkout, but it seems that this method cannot find branches from forks. It appears that `actions/checkout@v2` can automatically checkout?([*](https://github.com/WeblateOrg/weblate/issues/6240#issuecomment-869715886)) the source branch of the pull request on GitHub Actions, so it is necessary to delete the following sections that are being used:
  ```yaml
  with:
    ref: ${{ github.head_ref }}
  ```
  Specifically, these are:
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L23-L24
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L37-L38
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L50-L51
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L65-L66
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L93-L94
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-docker-build-check.yml#L16-L17
___

英語があまり得意な方ではないので、ミスリーディングを減らすために日本語も残しておきます。
以下の点ご確認いただけますと幸いです。

- 以下の`cloud2TransformedCloud()`では、`header.frame_id`(LiDARのframe)から`cam_model_.tfFrame()`(カメラのframe)に点群を変換する必要があります。  https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/src/tracker_with_cloud_node.cpp#L57

  一方で以下の`cloud2TransformedCloud()`では、`cam_model_.tfFrame()`(カメラのframe)から`header.frame_id`(LiDARのframe)に点群を変換する必要があります。  https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/src/tracker_with_cloud_node.cpp#L96

  しかしながら現在の実装では、上記の両方において`cloud2TransformedCloud()`が`cam_model_.tfFrame()`(カメラのframe)から`header.frame_id`(LiDARのframe)に点群を変換するようになってしまっています。

  ROS 2版ですが`cloud2TransformedCloud()`の実装例が以下にございますのでこちらを是非参考にしてください。  https://github.com/Alpaca-zip/ultralytics_ros/blob/a8a6a7e4d4a9a2609148f9dc8a6eed45d2ce9804/src/tracker_with_cloud_node.cpp#L245-L263

  **Notice**
  `lookupTransform()`の第一引数は`target_frame`、第二引数は`source_frame`であることにご注意ください。
  [Link to reference](https://docs.ros2.org/foxy/api/tf2_ros/classtf2__ros_1_1Buffer.html#a3ab502cc1e8b608957a96ad350815aee)

- 以下の箇所はプルリクエストのソースブランチを明示的に指定してチェックアウトすることを意味しますが、どうやらこの方法ではフォーク先のブランチは見つけられないようです。`actions/checkout@v2`はそれ単体でも自動でプルリクエストのソースブランチにチェックアウトをしてくれるらしい？([*](https://github.com/WeblateOrg/weblate/issues/6240#issuecomment-869715886))のでGithub Action上で利用している以下の箇所はすべて削除する必要があります。
  ```yaml
  with:
    ref: ${{ github.head_ref }}
  ```
  削除の対象は以下の全てです。
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L23-L24
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L37-L38
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L50-L51
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L65-L66
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-ci.yml#L93-L94
  - https://github.com/Alpaca-zip/ultralytics_ros/blob/41753e2308560601cfd88674032b5de8d1310e1c/.github/workflows/noetic-docker-build-check.yml#L16-L17

**Comment by h-wata** [comment]
@Alpaca-zip 
Thank you your replay and apologize my misunderstandings.
Made the fixes from the review. 
Could you please review the changes when you have a moment? Thanks.

File: src/tracker_with_cloud_node.cpp [hunk 1]
```diff
@@ -151,36 +155,28 @@ void TrackerWithCloudNode::processPointsWithMask(const pcl::PointCloud<pcl::Poin
...
+                                             const std::string& source_frame, const std::string& target_frame,
+                                             const ros::Time& stamp)
 {
   pcl::PointCloud<pcl::PointXYZ>::Ptr transformed_cloud(new pcl::PointCloud<pcl::PointXYZ>);
 
   try
   {
-    geometry_msgs::TransformStamped tf =
-        tf_buffer_->lookupTransform(header.frame_id, cam_model_.tfFrame(), header.stamp);
+    geometry_msgs::TransformStamped tf = tf_buffer_->lookupTransform(source_frame, target_frame, stamp);
```
Comment (Alpaca-zip) [inline src/tracker_with_cloud_node.cpp]: > Notice
> Please be aware that the first argument of `lookupTransform()` is `target_frame`, and the second argument is `source_frame`.
> [Link to reference](https://docs.ros2.org/foxy/api/tf2_ros/classtf2__ros_1_1Buffer.html#a3ab502cc1e8b608957a96ad350815aee)

```suggestion
    geometry_msgs::TransformStamped tf = tf_buffer_->lookupTransform(target_frame, source_frame, stamp);
```

**Review by Alpaca-zip — CHANGES_REQUESTED** [review]
> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844141/job/21508091103?pr=58

Please delete the following section.
https://github.com/Alpaca-zip/ultralytics_ros/blob/c23d77fde3fab75b5c0a10e8996f884b3dc5fe6d/.github/workflows/noetic-docker-build-check.yml#L16-L17
___

> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844137/job/21508091095?pr=58

> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844137/job/21508091683?pr=58

> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844137/job/21508091421?pr=58

> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844137/job/21508091860?pr=58

> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7881844137/job/21508091994?pr=58

I haven't encountered above issues with the GitHub Action before([*](https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7580983305)), and since it's not clear what's causing it, I'll handle it separately and keep an eye on things.

**Comment by Alpaca-zip** [comment]
> https://github.com/Alpaca-zip/ultralytics_ros/actions/runs/7883310972/job/21510311360?pr=58

Sorry for the inconvenience, it seems that the GitHub Action failed because the repository you forked to did not have access to Secrets. The Docker login is not required just to build the image, so could you please remove the code in the following section?
https://github.com/Alpaca-zip/ultralytics_ros/blob/dfbdb2351d81e53724e49568cc32368d2414d9c8/.github/workflows/noetic-docker-build-check.yml#L20-L24

I'd like to proceed with the merge after making this adjustment. Thank you for your cooperation.

**Comment by Alpaca-zip** [comment]
- [x] I have tested this change with KITTI data set.

**Review by Alpaca-zip — APPROVED** [review]

**Pull request merged**